mypy/plugin.py

```diff
@@ -379,3 +421,332 @@ def int_pow_callback(ctx: MethodContext) -> Type:
     return ctx.default_return_type
+
+
+# The names of the different functions that create classes or arguments.
+attr_class_makers = {
+    'attr.s',
+    'attr.attrs',
+    'attr.attributes',
+}
+attr_dataclass_makers = {
+    'attr.dataclass',
+}
+attr_attrib_makers = {
+    'attr.ib',
+    'attr.attrib',
+    'attr.attr',
+}
+
+
+class Attribute:
+    """The value of an attr.ib() call."""
+
+    def __init__(self, name: str, type: Optional[Type],
+                 has_default: bool, init: bool,
+                 context: Context) -> None:
+        self.name = name
+        self.type = type
+        self.has_default = has_default
+        self.init = init
+        self.context = context
+
+    def argument(self) -> Argument:
+        """Return this attribute as an argument to __init__."""
+        # Convert type not set to Any.
+        _type = self.type or AnyType(TypeOfAny.unannotated)
+        # Attrs removes leading underscores when creating the __init__ arguments.
+        return Argument(Var(self.name.strip("_"), _type), _type,
+                        None,
+                        ARG_OPT if self.has_default else ARG_POS)
+
+
+def attr_class_maker_callback(
+        attr_classes: Dict[TypeInfo, List[Attribute]],
+        ctx: ClassDefContext,
+        auto_attribs_default: bool = False
+) -> None:
+    """Add necessary dunder methods to classes decorated with attr.s.
+
+    Currently supports init=, cmp= and frozen= and auto_attribs=.
+    """
+    # attrs is a package that lets you define classes without writing dull boilerplate code.
+    #
+    # At a quick glance, the decorator searches the class body for assignments of `attr.ib`s (or
+    # annotated variables if auto_attribs=True), then depending on how the decorator is called,
+    # it will add an __init__ or all the __cmp__ methods.  For frozen=True it will turn the attrs
+    # into properties.
+    #
+    # See http://www.attrs.org/en/stable/how-does-it-work.html for information on how attrs works.
+
+    def called_function(expr: Expression) -> Optional[str]:
+        """Return the full name of the function being called by the expr, or None."""
+        if isinstance(expr, CallExpr) and isinstance(expr.callee, RefExpr):
+            return expr.callee.fullname
+        return None
+
+    def get_callable_type(call: CallExpr) -> Optional[CallableType]:
+        """Get the CallableType that's attached to a CallExpr."""
+        callable_type = None
+        if (isinstance(call, CallExpr)
+                and isinstance(call.callee, RefExpr)
+                and isinstance(call.callee.node, Var)
+                and call.callee.node.type):
+            callee_type = call.callee.node.type
+            if isinstance(callee_type, Overloaded):
+                # We take the last overload.
+                callable_type = callee_type.items()[-1]
+            elif isinstance(callee_type, CallableType):
+                callable_type = callee_type
+        return callable_type
```

Comment:
I was trying to programmatically extract the signature so that in the future I could support helper functions that return attribs.  (For example at my company we add a couple of arguments to do some special handling)  By looking at the stub rather than hardcoding the order and names, then this could be supported.

Reply:
Yes, I see. There is a trade-off between code clarity and flexibility. I will make another pass of review now to see how it reads.